ctp2_code/gfx/spritesys/director.cpp

```diff
@@ -667,126 +669,126 @@ class DirectorImpl : public Director {
 #endif
 
 	virtual void AddMove(
-		Unit                mover,
-		MapPoint const &    oldPos,
-		MapPoint const &    newPos,
-		sint32              numRevealed,
-		UnitActor **        revealedActors,
-		sint32              numRest,
-		UnitActor **        restOfStack,
-		sint32              soundID
+		const Unit     &mover,
+		MapPoint const &startPosition,
+		MapPoint const &endPosition,
+		sint32         numberOfRevealedActors,
+		UnitActor      **revealedActors,
+		sint32         numberOfMoveActors,
+		UnitActor      **moveActors,
+		sint32         soundID
 	);
 	virtual void AddTeleport(
-		Unit                top,
-		MapPoint const &    oldPos,
-		MapPoint const &    newPos,
-		sint32              numRevealed,
-		UnitActor **        revealedActors,
-		sint32              arraySize,
-		UnitActor **        moveActors
+		const Unit     &mover,
+		MapPoint const &startPosition,
+		MapPoint const &endPosition,
+		sint32         numberOfRevealedActors,
+		UnitActor      **revealedActors,
+		sint32         numberOfMoveActors,
+		UnitActor      **moveActors
 	);
-	virtual void AddAttack(Unit attacker, Unit attacked);
-	virtual void AddAttackPos(Unit attacker, MapPoint const & pos);
-	virtual void AddSpecialAttack(Unit attacker, Unit attacked, SPECATTACK attack);
-	virtual void AddDeath(UnitActor *dead, const MapPoint &deadPos, sint32 deadSoundID);
+	virtual void AddAttack(const Unit& attacker, const Unit& attacked);
+	virtual void AddAttackPos(const Unit& attacker, const MapPoint &position);
+	virtual void AddSpecialAttack(const Unit& attacker, const Unit& attacked, SPECATTACK attack);
+	virtual void AddDeath(UnitActor *dead, const MapPoint &deadPosition, sint32 deadSoundID);
 	virtual void AddProjectileAttack(
-		Unit shooting,
-		Unit target,
-		SpriteState *projectile_state,
-		SpriteState *projectileEnd_state,
-		sint32 projectile_Path
+		const Unit  &shooting,
+		const Unit  &target,
+		SpriteState *projectileState,
+		SpriteState *projectileEndState,
+		sint32      projectilePath
 	);
-	virtual void AddSpecialEffect(MapPoint &pos, sint32 spriteID, sint32 soundID);
-	virtual void AddMorphUnit(UnitActor *morphingActor, SpriteState *ss, sint32 type,  Unit id);
-	virtual void AddHide(Unit hider);
-	virtual void AddShow(Unit hider);
-	virtual void AddWork(Unit worker);
+	virtual void AddSpecialEffect(const MapPoint &position, sint32 spriteID, sint32 soundID);
+	virtual void AddMorphUnit(UnitActor *morphingActor, SpriteState *spriteState, sint32 type, const Unit &id);
+	virtual void AddHide(const Unit &hider);
```

Comment:
Why is this now a reference? This way, it should be now a 64 bit address on Linux and a 32 bit address on Windows. Even so the original object is just 32 bits.

In fact that argument can also be make for MapPoint, which are now 32 bits but were in CTP1 48 bits.

Reply:
Sorry for the long reply I want to explain and be concrete why I think this is a good change.

**Summary**: As in my opinion this is a technical improvement I propose to keep this change.

**Reasoning**
In general, when I am not optimizing, I do not want to know the exact size of all classes or the amount of time a copy-constructor takes when it is technically not needed. The exact size of an instance of a class may not be so obvious. The data-size may be small but in case a class has virtual methods also the virtual-table pointer will be part of the data and in case it is a derived class it becomes even more complicated. I am not 100% sure but depending on the situation there also may be a need for the `this` pointer to be stored.

In general, computers with registers have less work when using their native pointer-size as the values will be in registers that can be used immediately. In case a 16-bit value is stored in a 64-bit register it needs to be masked out (by shifting or by and-ing). As the compiler takes care of the pointer-size (ensuring that it fits the host) I, as developer, do not need to care about that which will ensure that I make less mistakes and be more efficient. 

So in case I need a class that is only used for reading some data without modifying the class and there is also no need for a copy (it will not be stored), I use the `const reference` as that gives me exactly what I need without any overhead. When an instance will be used in stead of a `const reference` the program will allocate space for the object (stack or heap) and execute a copy-constructor both are not needed in this situation.

BTW: Technically as the class already exists it may even be that the reference is given to the next method via a register which is not possible for the class itself which means less overhead. I am not sure but I also think that inlining a method becomes harder when the class itself is copied. While by using a `const reference` it is obvious for the compiler that it is the same class.

Reply:
I agree that this stuff is nothing we have to optimize, extensively. I just happened to have optimized some classes for size and also looked at the sizes of Unit and others to make sure that the Linux 64 bit executable can load game created by the Windows 32 bit executable.

So then we go here for consistency and ignore the implementation details, since they do not matter here. So we keep the change.